tl-its-umich-edu/peer-review-data
# get all review data related to the assignment (iss. #3)

Resolves #3.

## Run notes

For the one assignment used in development…
* Runtime: 1m30s–2m30s
* Number of objects
    * User: 382
    * Submission: 352 (Some users didn't make submissions)
    * Assessment: 1019 (Reviews, 2–3 per submission)
    * Criterion: 6
    * Comment: 6114 (6 criteria for 1019 assessments)

The app logs a **_LOT_** of `INFO`-level messages about what it's doing.  It sometimes helps to redirect the console output to a file to allow text searches.

## Test plan

1. Copy `config/.env.sample` to `.env` in the root of the project.  Edit the file to replace the values contained therein with valid ones for the test environment.  

    ```sh
    cp config/.env.sample .env
    vim .env
    ```
    * Use `canvas-test` for the host in the API URL
    * Use `545451` for the course ID
    * Use `1747535` for the assignment ID

1. Use Docker's `compose` tool to build the application, run it, and update the DB.

    ```sh
    docker compose up --build
    ```
    1. Development code is in place keeps the container running for four hours.  That allows connecting to the running container.

        ```sh
        docker compose exec -it main bash
        ```
    1. When connected to the running container, the application can be run again without restarting the entire container.

        ```sh
        python manage.py run
        ```
    1. After the container is stopped, if the environment needs to be cleared (for example, to test a first-time run again), use the commands…

        ```sh
        docker compose down -v
        rm -fr .data
        ```

1. To examine the data, connect to the DB and run SQL queries.  DB credentials can be found in the `db` service section of `docker-compose.yaml`.  

    A SQL query similar to that used to gather data for the MWrite dashboard (see [documentation](https://docs.google.com/document/d/1FEDjYng1FilD64xx_XHct1zff_Ds9T9hsswFw-p2_J4/edit#bookmark=id.1fob9te))…

    ```sql
    SELECT
      course.name AS Course,
      course.id AS CourseID,
      `assignment`.name AS Prompt,
      `assignment`.id AS PromptID,
      author.sortable_name AS Author,
      author.id AS AuthorID,
      reviewer.sortable_name AS Reviewer,
      reviewer.id AS ReviewerID,
      criterion.description AS Criterion,
      criterion.id AS CriterionID,
      comment.comments,
      comment.id AS CommentID,
      -- comment.commented_at_utc AS CommentTimeUTC
      TIMESTAMP('1970-01-01 00:00:01') AS CommentTimeUTC
    FROM
      comment
    INNER JOIN assessment ON
      comment.assessment_id = assessment.id
    INNER JOIN criterion ON
      comment.criterion_id = criterion.id
    INNER JOIN submission ON
      assessment.submission_id = submission.id
    INNER JOIN `assignment` ON
      submission.assignment_id = `assignment`.id
    INNER JOIN course ON
      `assignment`.course_id = course.id
    INNER JOIN `user` AS author ON
      submission.user_id = author.id
    INNER JOIN `user` AS reviewer ON
      assessment.assessor_id = reviewer.id
    ```

    Note that the data from Canvas does _**NOT**_  include timestamps on the review comments, so one cannot be provided here.  The UNIX epoch time is given here as a placeholder until Instructure can provide timestamps.
